ui/src/criteria/attribute.tsx

```diff
@@ -81,73 +136,216 @@ function AttributeEdit(props: AttributeEditProps) {
   const hintDisplayName = (hint: tanagra.EnumHintValue) =>
     hint.displayName || "Unknown Value";
 
-  const enumHintValues = underlay.entities
+  const attributeFilterHint = underlay.entities
     .find((g) => g.name === underlay.primaryEntity)
-    ?.attributes?.find((attribute) => attribute.name === props.data.attribute)
-    ?.attributeFilterHint?.enumHint?.enumHintValues;
+    ?.attributes?.find(
+      (attribute) => attribute.name === props.data.attribute
+    )?.attributeFilterHint;
+  const enumHintValues = attributeFilterHint?.enumHint?.enumHintValues;
+  const integerBoundsHint = attributeFilterHint?.integerBoundsHint;
+
+  const hasValidIntegerBounds =
+    isValid(integerBoundsHint?.min) && isValid(integerBoundsHint?.max);
+
+  if (hasValidIntegerBounds) {
+    const minBound = integerBoundsHint?.min || 0; // This is to ensure the compiler won't complain the object might be undefined.
+    const maxBound = integerBoundsHint?.max || 0; // although we already know that min and max is valid.
+
+    // two sets of values are needed due to the input box and slider is isolated.
+    const [minInputValue, setMinInputValue] = useState<number | string>(
+      props.data.min || minBound
+    );
+    const [maxInputValue, setMaxInputValue] = useState<number | string>(
+      props.data.max || maxBound
+    );
+
+    const updateValue = (newMin: number, newMax: number) => {
+      props.dispatchFn(
+        produce(props.data, (data) => {
+          data.min = newMin;
+          data.max = newMax;
+        })
+      );
+    };
+
+    const handleChange = (event: Event, newValue: number | number[]) => {
+      newValue = newValue as number[];
+      const [newMin, newMax] = newValue;
+      setMinInputValue(newMin);
+      setMaxInputValue(newMax);
+      updateValue(newMin, newMax);
+    };
+
+    const handleMinInputChange = (
+      event: React.ChangeEvent<HTMLInputElement>
+    ) => {
+      let newMin = event.target.value === "" ? "" : Number(event.target.value);
```

Comment:
The logic here can be cleaned up:
1. Call `setMinInputValue` as the first step.
2. Make `newMin` always a number: `const newMin = minInputValue === "" ? 0 || Number(minInputValue);`
3. You're using `maxValue` in multiple places, including above in `useState`. Pull it out into a constant.
4. Splitting the min and max across multiple lines is several of unnecessary `newMin`s. Merge it into one line: `updateValue(Math.min(maxValue, Math.max(minBound, newMin)), maxValue;`

Reply:
done